### PR TITLE
Search for discriminator children through multiple levels of allOf nesting

### LINKED ISSUE #1458: Data conversion error when using the azure_mgmt_recoveryservicesbackup crate
Using the `azure_mgmt_recoveryservicesbackup` version: `0.17.0` I'm receiving the following error when using the `backup_protectable_items_client` and calling the [list](https://learn.microsoft.com/en-us/rest/api/backup/backup-protectable-items/list?view=rest-backup-2023-02-01&tabs=HTTP) API endpoint:

> Error: Error { context: Custom(Custom { kind: DataConversion, error: Error("unknown variant `Microsoft.Compute/virtualMachines`, expected one of `AzureFileShare`, `AzureVmWorkloadProtectableItem`, `IaaSVMProtectableItem`", line:
 1, column: 761) }) }

The `enum` (`WorkloadProtectableItemUnion`)  in question can be found [here](https://github.com/Azure/azure-sdk-for-rust/blob/97d7e953a2f989702567504ca052d3bd114b973b/services/mgmt/recoveryservicesbackup/src/package_2023_04/models.rs#L11900C2-L11900C2).

I can see the following OpenAPI definitions for the 3 expected types in the error:

- [AzureFileShare](https://github.com/Azure/azure-rest-api-specs/blob/95c4286028ad9fecb3227288ff7345e528e7890d/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-02-01/bms.json#L8502)
- [AzureVmWorkloadProtectableItem](https://github.com/Azure/azure-rest-api-specs/blob/95c4286028ad9fecb3227288ff7345e528e7890d/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-02-01/bms.json#L8708)
- [IaaSVMProtectableItem](https://github.com/Azure/azure-rest-api-specs/blob/95c4286028ad9fecb3227288ff7345e528e7890d/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-02-01/bms.json#L10312C14-L10312C14)

This looks like an issue with their OpenAPI spec as I don't see a model definition with the `Microsoft.Compute/virtualMachines` `discriminator-value`.

Before I go posting an issue on their repo I wanted to confirm that the issue is indeed their specification file and not an issue with the auto generated SDK.

### PR DESCRIPTION
**Pull request by krishanjmistry**
Attempts to fix #1458 - see the issue for more details.

I've tested this by modifying the `codegen/examples/resources_mgmt.rs` to point to:
`azure-rest-api-specs/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json`

This now produces the enum like so:
```rust
#[doc = "Type of the backup item."]
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "protectableItemType")]
pub enum WorkloadProtectableItemUnion {
    AzureFileShare(AzureFileShareProtectableItem),
    #[serde(rename = "Microsoft.ClassicCompute/virtualMachines")]
    MicrosoftClassicComputeVirtualMachines(AzureIaaSClassicComputeVmProtectableItem),
    #[serde(rename = "Microsoft.Compute/virtualMachines")]
    MicrosoftComputeVirtualMachines(AzureIaaSComputeVmProtectableItem),
    AzureVmWorkloadProtectableItem(AzureVmWorkloadProtectableItem),
    #[serde(rename = "SAPAseSystem")]
    SapAseSystem(AzureVmWorkloadSapAseSystemProtectableItem),
    #[serde(rename = "SAPHanaDBInstance")]
    SapHanaDbInstance(AzureVmWorkloadSapHanaDbInstance),
    #[serde(rename = "SAPHanaDatabase")]
    SapHanaDatabase(AzureVmWorkloadSapHanaDatabaseProtectableItem),
    #[serde(rename = "HanaHSRContainer")]
    HanaHsrContainer(AzureVmWorkloadSapHanaHsrProtectableItem),
    #[serde(rename = "SAPHanaSystem")]
    SapHanaSystem(AzureVmWorkloadSapHanaSystemProtectableItem),
    #[serde(rename = "SQLAvailabilityGroupContainer")]
    SqlAvailabilityGroupContainer(AzureVmWorkloadSqlAvailabilityGroupProtectableItem),
    #[serde(rename = "SQLDataBase")]
    SqlDataBase(AzureVmWorkloadSqlDatabaseProtectableItem),
    #[serde(rename = "SQLInstance")]
    SqlInstance(AzureVmWorkloadSqlInstanceProtectableItem),
    #[serde(rename = "IaaSVMProtectableItem")]
    IaaSvmProtectableItem(IaaSvmProtectableItem),
}
```

I'm unsure how to test this more generally across all the API's - happy to be steered on this

### CODE REVIEW

**Comment by krishanjmistry** [comment]
I'm starting to think it should be a breadth-first search rather than depth-first (what I've currently implemented). That way, we'd stop once we first hit a discriminator value...

**Comment by cataggar** [comment]
Thanks @krishanjmistry! I ran the code generation. Here is how I ran it:

``` nushell
~/ms/azure-rest-api-specs> git checkout 7c906b15744aaf7b782dfb82ccefcbef05cbf20b
~/ms/azure-sdk-for-rust/services/autorust> cargo run -p azure-autorust --release
```

It fixed 7 control plane and 3 data place services.

Once a unit test is added, this should be good to merge.

**Review by cataggar — APPROVED** [review]
Thanks again! For extra credit, a unit test that used to fail deserialization, but now works would be nice. json from a specification example could probably be copied.

**Comment by krishanjmistry** [comment]
I've added a test in the `mgmt/recoveryservicesbackup` crate which does just this. Based off of the example in REST API specs which handily used one of the previous problem cases

File: services/mgmt/recoveryservicesbackup/tests/nested_discriminator_value.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/// This test is added to verify that the [WorkloadProtectableItemUnion] contains
+/// discriminator values that indirectly are discriminator children of this discriminator
+fn test_deserialization_of_nested_all_of_property() -> Result<()> {
+    let bytes = read("tests/backup_protectable_item_list.json")?;
+
+    let list: WorkloadProtectableItemResourceList = serde_json::from_slice(&bytes)?;
+
+    let item = list.value.first().expect("There should be at least one item in the list");
+    let properties= item.properties.as_ref().expect("The properties should be present");
+    assert!(matches!(properties, &WorkloadProtectableItemUnion::MicrosoftClassicComputeVirtualMachines(_)));
```
Comment (cataggar) [inline services/mgmt/recoveryservicesbackup/tests/nested_discriminator_value.rs]: Perfect. Thanks!

**Pull request merged**